packager/mux/mux.go

```diff
@@ -37,82 +37,134 @@ import (
 	sLen := len(s)
 	trimmed := strings.TrimPrefix(s, prefix)
-	return trimmed, len(trimmed) != sLen
+	return trimmed, len(prefix)+len(trimmed) == sLen
 }
 
 var allowedMethods = map[string]bool{http.MethodGet: true, http.MethodHead: true}
 
 func (this *mux) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
-	if !allowedMethods[req.Method] {
-		http.Error(resp, "405 method not allowed", http.StatusMethodNotAllowed)
-		return
-	}
-
-	params := map[string]string{}
-	req = WithParams(req, params)
-
-	// Use EscapedPath rather than RequestURI because the latter can take
-	// absolute-form, per https://tools.ietf.org/html/rfc7230#section-5.3.
-	//
-	// Use EscapedPath rather than RawPath to verify that the request URI
-	// is a valid encoding, per
-	// https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange
-	// item 3.
 	path := req.URL.EscapedPath()
-	if suffix, ok := tryTrimPrefix(path, "/priv/doc"); ok {
-		if suffix == "" {
-			this.signer.ServeHTTP(resp, req)
-		} else if suffix[0] == '/' {
-			params["signURL"] = suffix[1:]
-			if req.URL.RawQuery != "" {
-				params["signURL"] += "?" + req.URL.RawQuery
-			}
-			this.signer.ServeHTTP(resp, req)
-		} else {
-			http.NotFound(resp, req)
+
+	// Find the first matching routing rule.
+	matchingRule := (*routingRule)(nil)
+	var suffix string
+	for _, currentRule := range this.routingMatrix {
+		if currentSuffix, isMatch := tryTrimPrefix(path, currentRule.URLPathPrefix); isMatch {
+			matchingRule = &currentRule
+			suffix = currentSuffix
+			break
 		}
-	} else if suffix, ok := tryTrimPrefix(path, util.CertURLPrefix+"/"); ok {
-		unescaped, err := url.PathUnescape(suffix)
-		if err != nil {
-			http.Error(resp, "400 bad request - bad URL encoding", http.StatusBadRequest)
+	}
+
+	errorMsg := ""
+	errorCode := 0
+	if matchingRule == nil {
+		// Should never happen, because the last rule matches any path.
+		errorMsg, errorCode = "500 internal error", http.StatusInternalServerError
+	} else {
+		// Validate HTTP method and params, parse params and attach them to req.
+		if !allowedMethods[req.Method] {
+			errorMsg, errorCode = "405 method not allowed", http.StatusMethodNotAllowed
 		} else {
-			params["certName"] = unescaped
-			this.certCache.ServeHTTP(resp, req)
+			params := map[string]string{}
+			req = WithParams(req, params)
+			matchingRule.suffixValidatorFunc(suffix, req, &params, &errorMsg, &errorCode)
 		}
-	} else if path == util.HealthzPath {
-		this.healthz.ServeHTTP(resp, req)
-	} else if path == util.ValidityMapPath {
-		this.validityMap.ServeHTTP(resp, req)
+	}
+
+	// Prepare the handler.
+	var handlerFunc http.Handler
+	if errorCode != 0 {
+		handlerFunc = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { http.Error(w, errorMsg, errorCode) })
 	} else {
-		http.NotFound(resp, req)
+		handlerFunc = matchingRule.handler
 	}
+
+	handlerFunc.ServeHTTP(resp, req)
```

Comment:
Optionally, write this as:

```
if errorCode == 0 {
  matchingRule.handler.ServeHTTP(resp, req)
} else {
  http.Error(w, errorMsg, errorCode)
}
```

Benefits (IMO):
 1. Gets rid of a double-negative (the `else` of a `!=`).
 2. Inlines an abstraction that's only used once (`handlerFunc`).
 3. More concise.

Reply:
I see you met in the middle (did 1 but not 2). That's fine by me, but I'm curious why (I'm betting you thought about it more than me, and I like learning things).